src/porepy/grids/mortar_grid.py

```diff
@@ -273,46 +306,65 @@ def update_secondary(self, new_g: pp.Grid, tol: float = None) -> None:
+                """update_secondary() is only implemented when the secondary
                 grid has the same dimension as the mortar grid"""
             )
 
-        split_matrix = {}
+        # Build mappings for integrated and averaged quantities separately.
+        split_matrix_avg, split_matrix_int = {}, {}
 
         # For each side we compute the mapping between the new lower dimensional
-        # grid and the mortar grid, we store them in a dictionary with SideTag as key.
+        # grid and the mortar grid, we store them in a dictionary with MortarSide as key.
+        # IMPLEMENTATION NOTE: The loop, and some of the complexity below, is necessary
+        # to allow for different grids on the mortar sides (if more than one). It is
+        # not clear that our full implementation supports this (no obvious weak points),
+        # but AFAIK, it has never been tested, but we will keep the generality for now.
         for side, g in self.side_grids.items():
             if g.dim != new_g.dim:
                 raise ValueError("Grid dimension has to be the same")
 
             if self.dim == 0:
-                # Nothing to do
+                # Nothing to do, all 0d grids are matching
                 return
             elif self.dim == 1:
-                split_matrix[side] = _split_matrix_1d(g, new_g, tol).T
+                # We need to map from the new to the old grid.
+                # See below for ideas of how to allow for updating the secondary grid
+                # more than once.
+                mat_avg = pp.match_grids.match_1d(g, new_g, tol, scaling="averaged")
+                mat_int = pp.match_grids.match_1d(g, new_g, tol, scaling="integrated")
             elif self.dim == 2:
-                split_matrix[side] = _split_matrix_2d(g, new_g, tol).T
+                mat_avg = pp.match_grids.match_2d(g, new_g, tol, scaling="averaged")
+                mat_int = pp.match_grids.match_2d(g, new_g, tol, scaling="integrated")
             else:
                 # No 3d mortar grid
                 raise ValueError
 
+            split_matrix_avg[side] = mat_avg
+            split_matrix_int[side] = mat_int
         # In the case of different side ordering between the input data and the
         # stored we need to remap it. The resulting matrix will be a block
         # matrix, where in each block we have the mapping between the
         # (relative to side) the new grid and the mortar grid.
-        matrix = np.empty((self.num_sides(), 1), dtype=object)
+        matrix_avg = np.empty((self.num_sides(), 1), dtype=object)
+        matrix_int = np.empty((self.num_sides(), 1), dtype=object)
 
         for pos, (side, _) in enumerate(self.side_grids.items()):
-            matrix[pos, 0] = split_matrix[side]
+            matrix_avg[pos] = split_matrix_avg[side]
+            matrix_int[pos] = split_matrix_int[side]
+
+        # IMPLEMENTATION NOTE: To allow for replacing the secondary grid more than once,
```

Comment:
Good, somebody might be very thankful for this in the future!

Reply:
That somebody may very well be me.